File: Part2/Java/Tree/_6_All_traversals.java
package Java.Tree;

import java.io.*;
import java.util.*;

class TreeNode_6_{
    public int data;
    public TreeNode_6_ left;
    public TreeNode_6_ right;
    public TreeNode_6_(){}
    public TreeNode_6_(int val){
        this.data=val;
        this.left=null;
        this.right=null;
    }
}

class Pair_<T, U> {
    public final T first;
    public final U second;

    public Pair_(T first, U second) {
        this.first = first;
        this.second = second;
    }
}

public class _6_All_traversals {
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(InputStream inputStream) {
            br = new BufferedReader(new InputStreamReader(inputStream));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

    /*
    Tree - non linear data structure
    */


    // Tree Traversals

    // TC:O(N) SC:O(N) 
    public static void All_In_One_Traversals(TreeNode_6_ root){
        ArrayList<Integer> preorder = new ArrayList<>();
        ArrayList<Integer> inorder = new ArrayList<>();
        ArrayList<Integer> postorder = new ArrayList<>();

        Stack<Pair_<TreeNode_6_,String>> st = new Stack<>(); // stack 1

        // st.push(new Pair_<TreeNode_6_, String>(root, "preorder"));
        st.push(new Pair_<>(root,"preorder"));
        
        while(!st.isEmpty()){
            Pair_<TreeNode_6_, String> n = st.peek();
            st.pop();

            if(n.second == "preorder"){
                preorder.add(n.first.data);
                st.push(new Pair_<>(n.first, "inorder"));
                if(n.first.left!=null){
                    st.push(new Pair_<>(n.first.left, "preorder"));
                }
            }
            else if(n.second == "inorder"){
                inorder.add(n.first.data);
                st.push(new Pair_<>(n.first,"postorder"));
                if(n.first.right!=null){
                    st.push(new Pair_<>(n.first.right, "preorder"));
                }
            }
            else if(n.second == "postorder"){
                postorder.add(n.first.data);
            }
        }
        for(int i=0;i<inorder.size();i++){
            System.out.println(preorder.get(i)+" "+inorder.get(i)+" "+postorder.get(i));
        }
    }

    public static void solve(FastReader sc) {
        TreeNode_6_ root = new TreeNode_6_(1);
        TreeNode_6_ two = new TreeNode_6_(2);
        TreeNode_6_ three = new TreeNode_6_(3);
        TreeNode_6_ four = new TreeNode_6_(4);
        TreeNode_6_ five = new TreeNode_6_(5);
        TreeNode_6_ six = new TreeNode_6_(6);
        TreeNode_6_ seven = new TreeNode_6_(7);

        root.left = two;
        root.right = five;

        two.left=three;
        two.right=four;

        five.left=six;
        five.right=seven;

        All_In_One_Traversals(root);
    }


    public static void main(String[] args) {
        try {
            System.setIn(new FileInputStream("input.txt"));
            System.setOut(new PrintStream(new FileOutputStream("output.txt")));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        FastReader sc = new FastReader(System.in);
        // int test = sc.nextInt();
        // while (test-- > 0) {
        // solve(sc);
        // }
        solve(sc);
    }
}

File: Part2/Part_B/_5_Trees/Practice /_6_All_traversals.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define mod 1000000007

/*
Tree - non linear data structure
*/

struct Node{
    int data;
    struct Node* left;
    struct Node* right;
    Node(){}
    Node(int val){
        data = val;
        left = nullptr;
        right = nullptr;
    }
}typedef Node;

class Solution{
public:

    // TC:O(n) SC:O(n) + O(n)
    void AllTraversals(Node* root){
        vector<int> preorder,inorder,postorder;

        stack<pair<Node*,string>> st;
        // add root initially 
        st.push({root, "preorder"});

        // preorder - 1
        // inorder - 2
        // postorder - 3
        while(!st.empty()){
            pair<Node*,string> n = st.top();
            st.pop();

            if(n.second == "preorder"){
                preorder.push_back(n.first->data);
                st.push({n.first,"inorder"});
                if(n.first->left!=nullptr){
                    st.push({n.first->left,"preorder"});
                }
            }
            else if(n.second == "inorder"){
                inorder.push_back(n.first->data);
                st.push({n.first,"postorder"});
                if(n.first->right!=nullptr){
                    st.push({n.first->right,"preorder"});
                }
            }
            else if(n.second == "postorder"){
                postorder.push_back(n.first->data);
            }
        }
        for(int i=0;i<preorder.size();i++){
            cout<<preorder[i]<<" "<<inorder[i]<<" "<<postorder[i]<<endl;
        }
    }
};

/*
        1
       /  \
      2    3
        \
        4

*/
void solve(){
    Node root(1);
    Node two(2);
    Node three(3);
    Node four(4);
    Node five(5);
    Node six(6);
    Node seven(7);
    Node eight(8);

    root.left = &two;
    root.right = &seven;

    two.left=&three;
    two.right=&four;
    four.left=&five;
    four.right=&six;

    Solution s;
    s.AllTraversals(&root);
}


int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
#ifndef ONLINE_JUDGE
    freopen("../../input.txt", "r", stdin);
    freopen("../../output.txt", "w", stdout);
#endif

    // ll test;
    // cin>>test;
    // while(test--)
    // {
    //     solve();
    // }
    solve();
    return 0;
}


